packages/react-devtools-shared/src/devtools/views/Components/ComponentsResizer.js

```diff
@@ -0,0 +1,209 @@
+import type {ElementRef} from 'react';
+
+import * as React from 'react';
+import {useEffect, useLayoutEffect, useReducer, useRef} from 'react';
+import {
+  localStorageGetItem,
+  localStorageSetItem,
+} from 'react-devtools-shared/src/storage';
+import styles from './ComponentsResizer.css';
+
+const LOCAL_STORAGE_KEY = 'React::DevTools::createResizeReducer';
+const VERTICAL_MODE_MAX_WIDTH = 600;
+const MINIMUM_SIZE = 50;
+
+type Props = {|
+  children: ({
+    resizeElementRef: ElementRef<HTMLElement>,
+    onResizeStart: () => void,
+  }) => React$Node,
+|};
+
+export default function ComponentsResizer({children}: Props) {
+  const wrapperElementRef = useRef(null);
+  const resizeElementRef = useRef(null);
+  const [state, dispatch] = createResizeReducer(
+    wrapperElementRef,
+    resizeElementRef,
+  );
+
+  const {isResizing} = state;
+
+  const onResizeStart = () =>
+    dispatch({type: 'ACTION_SET_IS_RESIZING', payload: true});
+  const onResizeEnd = () =>
+    dispatch({type: 'ACTION_SET_IS_RESIZING', payload: false});
+
+  const onResize = event => {
+    const resizeElement = resizeElementRef.current;
+    const wrapperElement = wrapperElementRef.current;
+
+    if (!isResizing || wrapperElement === null || resizeElement === null) {
+      return;
+    }
+
+    event.preventDefault();
+
+    const orientation = getOrientation(wrapperElementRef);
+
+    const {height, width, left, top} = wrapperElement.getBoundingClientRect();
+
+    const currentMousePosition =
+      orientation === 'horizontal' ? event.clientX - left : event.clientY - top;
+
+    const boundaryMin = MINIMUM_SIZE;
+    const boundaryMax =
+      orientation === 'horizontal'
+        ? width - MINIMUM_SIZE
+        : height - MINIMUM_SIZE;
+
+    const isMousePositionInBounds =
+      currentMousePosition > boundaryMin && currentMousePosition < boundaryMax;
+
+    if (isMousePositionInBounds) {
+      const resizedElementDimension =
+        orientation === 'horizontal' ? width : height;
+      const actionType =
+        orientation === 'horizontal'
+          ? 'ACTION_SET_HORIZONTAL_PERCENTAGE'
+          : 'ACTION_SET_VERTICAL_PERCENTAGE';
+      const percentage = (currentMousePosition / resizedElementDimension) * 100;
+
+      resizeElement.style.setProperty(
+        `--${orientation}-resize-percentage`,
+        `${percentage}%`,
+      );
```

Comment:
We are now setting a css variable based on the orientation.

Possible values:
`--horizontal-resize-percentage`
`--vertical-resize-percentage`

We use these two variables in CSS to determine the correct resize percentage based on the orientation using the media query that we have set up there.

I initially thought of using a ResizeObserver but it's far too complex, I think.

Thoughts, @bvaughn ?